File: rental/src/main/java/io/axoniq/demo/bikerental/rental/ui/Simulator.java
package io.axoniq.demo.bikerental.rental.ui;

import io.axoniq.demo.bikerental.coreapi.rental.CountOfBikesByTypeQuery;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import io.axoniq.demo.bikerental.rental.support.CommandDispatcher;
import io.axoniq.demo.bikerental.rental.support.QueryDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.util.UUID;
import java.util.concurrent.ExecutionException;

@Component
@Profile("simulator")
@EnableAsync
public class Simulator {

    private final CommandDispatcher commandDispatcher;
    private final QueryDispatcher queryDispatcher;
    private final BikeRentalDataGenerator bikeRentalDataGenerator;

    @Value("${inventory.size}")
    private int inventorySize;

    @Value("${inventory.bikeType}")
    private String inventoryBikeType;

    @Value("${rentalSimulation.bikeType}")
    private String rentalBikeType;

    @Value("${rentalSimulation.loops}")
    private int loops;

    @Value("${rentalSimulation.concurrency}")
    private int concurrency;

    @Value("${rentalSimulation.abandonPaymentFactor}")
    private int abandonPaymentFactor;

    @Value("${rentalSimulation.delayBetweenLoops}")
    private int delayBetweenLoops;

    Logger logger = LoggerFactory.getLogger(Simulator.class);

    public Simulator(CommandDispatcher commandDispatcher,
                     QueryDispatcher queryDispatcher,
                     BikeRentalDataGenerator bikeRentalDataGenerator) {
        this.commandDispatcher = commandDispatcher;
        this.queryDispatcher = queryDispatcher;
        this.bikeRentalDataGenerator = bikeRentalDataGenerator;
    }

    public void updateInventoryCreationConfiguration(int sizeOfBikeInventory, String bikeType) {
        this.inventorySize = sizeOfBikeInventory;
        this.inventoryBikeType = bikeType;
    }

    public void updateRentalGenerationConfiguration(String rentalBikeType, int loops, int concurrency, int abandonPaymentFactor, int delay) {
        this.rentalBikeType = rentalBikeType;
        this.loops = loops;
        this.concurrency = concurrency;
        this.abandonPaymentFactor = abandonPaymentFactor;
        this.delayBetweenLoops = delay;
    }

    @Scheduled(fixedRate = 25000, initialDelay = 5000)
    private void generateData() {
        try {
            this.generateBikes();
        } catch (Exception ex) {
            logger.error("error generating inventory", ex);
        }

        this.generateRentals();
    }

    public void generateBikes() throws ExecutionException, InterruptedException {
        long currentBikeCount = queryDispatcher.query(CountOfBikesByTypeQuery.class.getName(),
                                                      new CountOfBikesByTypeQuery(this.inventoryBikeType),
                                                      Long.class).get();

        if (currentBikeCount < this.inventorySize) {
            for (int i = 0; i < 10; i++) {
                commandDispatcher.send(new RegisterBikeCommand(UUID.randomUUID().toString(),
                                                               this.inventoryBikeType,
                                                               this.bikeRentalDataGenerator.randomLocation()));
            }
        }
    }

    private void generateRentals() {
        this.bikeRentalDataGenerator.generateRentals(this.rentalBikeType,
                                                     this.loops,
                                                     this.concurrency,
                                                     this.abandonPaymentFactor,
                                                     this.delayBetweenLoops).subscribe(logger::info);
    }
}

File: rental/src/ui/simulator.ts
import { randomUUID } from 'node:crypto';
import {
  CommandDispatcher,
  CountOfBikesByTypeQuery,
  createLogger,
  JavaTypes,
  QueryDispatcher,
  RegisterBikeCommand,
} from '@bikerental/core-api';
import type { config } from '../config.js';
import { BikeRentalDataGenerator } from './bike-rental-data-generator.js';

const logger = createLogger('Simulator');

const FIXED_RATE_MS = 25_000;
const INITIAL_DELAY_MS = 5_000;

/**
 * The 'simulator' profile component: every 25s it tops up the bike inventory and kicks off a batch of
 * rental cycles. Active only when the simulator profile is enabled (see {@link config}).
 */
export class Simulator {
  private inventorySize: number;
  private inventoryBikeType: string;
  private rentalBikeType: string;
  private loops: number;
  private concurrency: number;
  private abandonPaymentFactor: number;
  private delayBetweenLoops: number;
  private timer?: NodeJS.Timeout;

  constructor(
    private readonly commandDispatcher: CommandDispatcher,
    private readonly queryDispatcher: QueryDispatcher,
    private readonly dataGenerator: BikeRentalDataGenerator,
    settings: (typeof config)['simulator'],
  ) {
    this.inventorySize = settings.inventorySize;
    this.inventoryBikeType = settings.inventoryBikeType;
    this.rentalBikeType = settings.rentalBikeType;
    this.loops = settings.loops;
    this.concurrency = settings.concurrency;
    this.abandonPaymentFactor = settings.abandonPaymentFactor;
    this.delayBetweenLoops = settings.delayBetweenLoops;
  }

  start(): void {
    setTimeout(() => {
      void this.generateData();
      this.timer = setInterval(() => void this.generateData(), FIXED_RATE_MS);
    }, INITIAL_DELAY_MS);
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
    }
  }

  updateInventoryCreationConfiguration(sizeOfBikeInventory: number, bikeType: string): void {
    this.inventorySize = sizeOfBikeInventory;
    this.inventoryBikeType = bikeType;
  }

  updateRentalGenerationConfiguration(
    rentalBikeType: string,
    loops: number,
    concurrency: number,
    abandonPaymentFactor: number,
    delay: number,
  ): void {
    this.rentalBikeType = rentalBikeType;
    this.loops = loops;
    this.concurrency = concurrency;
    this.abandonPaymentFactor = abandonPaymentFactor;
    this.delayBetweenLoops = delay;
  }

  private async generateData(): Promise<void> {
    try {
      await this.generateBikes();
    } catch (error) {
      logger.error('error generating inventory', error);
    }
    void this.generateRentals();
  }

  private async generateBikes(): Promise<void> {
    const currentBikeCount =
      (await this.queryDispatcher.query<number>(
        CountOfBikesByTypeQuery.TYPE,
        new CountOfBikesByTypeQuery(this.inventoryBikeType),
        JavaTypes.LONG,
      )) ?? 0;

    if (currentBikeCount < this.inventorySize) {
      for (let i = 0; i < 10; i++) {
        void this.commandDispatcher.send(
          new RegisterBikeCommand(randomUUID(), this.inventoryBikeType, this.dataGenerator.randomLocation()),
        );
      }
    }
  }

  private generateRentals(): Promise<void> {
    return this.dataGenerator.generateRentals(
      this.rentalBikeType,
      this.loops,
      this.concurrency,
      this.abandonPaymentFactor,
      this.delayBetweenLoops,
      (line) => logger.info(line.trim()),
    );
  }
}
